File: python/reverseLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def createList(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        last_node = self.head
        while last_node.next:
            last_node = last_node.next
        last_node.next = new_node
        
    def reverseLinkedList(self):
        prev = None
        current = self.head
        while current:
            next_node = current.next
            current.next = prev
            prev = current

            current = next_node
        self.head = prev

    def printList(self):
        temp = self.head
        while temp:
            print(temp.data, end=" -> ")
            temp = temp.next

        
def main():
    list = LinkedList()
    list.createList(10)
    list.createList(20)
    list.createList(30)
    list.createList(40)
    print("list before reversing")
    list.printList()

    print("printing reverse liked list")
    list.reverseLinkedList()
    list.printList()

if __name__=="__main__":
    main()
    


    

File: javaFile/reverse_linked_list.java
class Node {
    int data;
    Node next;
    Node(int data) {
        this.data = data;
        this.next = null;
    }
    
}

class LinkedList {
    Node head;
    LinkedList() {
        this.head = null;
    }

    void create_list(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            return;
        }
        Node last_node = newNode;
        while (last_node != null) {
            last_node = last_node.next;
            
        }
        last_node.next = newNode;
        
    }

    void reverseLikedList() {
        Node prev = null;
        Node current = head;
        while ( current != null) {
            Node nextNode = current.next;
            current.next = prev;
            prev = current;
            
            current = nextNode;
            
        }
        head = prev;

    }
    void printList() {
        Node temp = head;
        while ( temp != null) {
            System.out.println(temp.data+"->");
            temp = temp.next;
            
        }
        System.out.println("none");
    }
}

public class reverse_linked_list {

     public static void main(String[] args) {
        LinkedList list = new LinkedList();
        list.create_list(10);
        list.create_list(20);
        list.create_list(30);
        list.create_list(40);
        System.out.println("printing list befor reverse");
        
        list.printList();
        list.reverseLikedList();
        System.out.println("printing list after reversing");
        list.printList();

     }
}